### PR TITLE
Add a lots of missing comments in ORCA

### PR DESCRIPTION
**Pull request by hardikar**
See each individual commit for the topic.

### CODE REVIEW

File: libgpopt/include/gpopt/base/CCTEInfo.h [hunk 1]
```diff
@@ -36,7 +36,10 @@ namespace gpopt
 	//		CCTEInfo
 	//
 	//	@doc:
-	//		Global CTE information
+	//		Global information about common table expressions (CTEs) including:
+	//		- the expression tree that defines each CTE
+	//		- the number of consumers created the optimizer
+	//		- a mapping from consumer columns to producer columns
```
Comment (dhanashreek89) [inline libgpopt/include/gpopt/base/CCTEInfo.h]: the number of consumers created **by** the optimizer ?

**Comment by hsyuan** [comment]
+1.
Nice work.

File: libgpopt/include/gpopt/base/CColumnFactory.h [hunk 2]
```diff
@@ -36,8 +36,13 @@ namespace gpopt
 	//		CColumnFactory
 	//
 	//	@doc:
-	//		Manager of column references; owns the memory in which they are
-	//		allocated, provides factory methods to generate new column references
+	//		Singleton factory class used to generate and manage CColRefs in ORCA.
+	//		The create CColRef objects are maintained in a hash table keyed by
```
Comment (bhuvnesh2703) [inline libgpopt/include/gpopt/base/CColumnFactory.h]: created

File: libgpopt/include/gpopt/base/CQueryContext.h [hunk 3]
```diff
@@ -28,7 +28,22 @@ namespace gpopt
...
 	//
 	//	@doc:
-	//		Query context object
+	//		Query specific information that optimizer receives as input
+	//		representing the requirements that need to be satisfied by the final
+	//		plan. This includes:
+	//		- Input logical expression
+	//		- Required columns
+	//		- Required plan (physical) properties at the top level of the query.
+	//		  This will include sort order, rewindability etc requested by the entire
```
Comment (bhuvnesh2703) [inline libgpopt/include/gpopt/base/CQueryContext.h]: may be you can add, that partition propagation spec is empty at this point, as it cannot be determined here. It's later initialized in Init Part Propagation function call.

File: libgpopt/include/gpopt/search/CJob.h [hunk 4]
```diff
@@ -31,6 +31,52 @@ namespace gpopt
...
+	//		A job can have any number of dependent (child) jobs. Execution within a
+	//		job cannot proceed as long as there is one or more dependent jobs that
+	//		are not finished yet.  Pausing a child job does not also allow the
+	//		parent job to proceed. The number of job dependencies (children) is
+	//		maintained by the member variable CJob::m_ulpRefs. The increment and
+	//		decrement of number of children are atomic operations performed by
+	//		CJob::IncRefs() and CJob::UlpDecrRefs() functions, respectively.
+	//
+	//		Job Queue:
+	//		Each job maintains a job queue CJob::m_pjq of other identical jobs that
```
Comment (bhuvnesh2703) [inline libgpopt/include/gpopt/search/CJob.h]: do we need to add a comment the capability of performing parallel task is concurrently not used, and its single threaded?

File: libgpopt/include/gpopt/search/CJob.h [hunk 5]
```diff
@@ -31,6 +31,52 @@ namespace gpopt
...
+	//		Job reentrance:
+	//		All optimization jobs are designed to be reentrant. This means that
+	//		there is a mechanism to support pausing a given job J1, moving
+	//		execution to another job J2, and then possibly returning to J1 to
+	//		resume execution. The re-entry point to J1 must be the same as the
+	//		point where J1 was paused. This mechanism is implemented using a state
+	//		machine.
+	//
+	//		Job Execution:
+	//		Each job defines two enumerations EState to define the different states
```
Comment (bhuvnesh2703) [inline libgpopt/include/gpopt/search/CJob.h]: enumerations,

File: libgpopt/include/gpopt/search/CJobGroupExpressionOptimization.h [hunk 6]
```diff
@@ -31,6 +31,12 @@ namespace gpopt
 	//	@doc:
 	//		Group expression optimization job
 	//
+	//		Responsible for finding the best plan rooted by a given group
```
Comment (bhuvnesh2703) [inline libgpopt/include/gpopt/search/CJobGroupExpressionOptimization.h]: Do you think this should be rephrased, this is technically not for finding the best plan, its for finding all the equivalent physical plans which could provide the required properties (and ensuring during the process physical properties are enforced), the plan which fulfills the required properties and has the lowest cost is filtered later. PgexprBestplan does finding the best plan?

File: libgpopt/src/engine/CEngine.cpp [hunk 7]
```diff
@@ -2140,8 +2146,38 @@ CEngine::SamplePlans()
...
+//		CEnfdProp::AppendEnforcers() is called for each of the enforced properties.
+//
+//		Returns true if no enforcers were created because they were deemed
+//		unnecessary or optional i.e all enforced properties were satisfied for
+//		the group expression under the current optimization context.  Returns
+//		false otherwise.
+//
+//		NB: This method is only concerned with a certain enforcer needs to be
+//		added into the group. Once added, there is no connection between the
+//		enforcer and the operator that created it. That is although sone group
```
Comment (asubramanya) [inline libgpopt/src/engine/CEngine.cpp]: s/sone/some/

File: libgpopt/src/engine/CEngine.cpp [hunk 8]
```diff
@@ -2140,8 +2146,38 @@ CEngine::SamplePlans()
...
+//		expression X createed the enforcer E, later, during costing, E can still
```
Comment (asubramanya) [inline libgpopt/src/engine/CEngine.cpp]: s/createed/created/

File: libgpopt/src/engine/CEngine.cpp [hunk 9]
```diff
@@ -2382,9 +2430,17 @@ CEngine::FCheckReqdPartPropagation
 //		CEngine::FCheckReqdProps
 //
 //	@doc:
-//		Check required properties;
-//		return false if it's impossible for the operator to satisfy one or more;
+//		Determing if checking required properties is needed.
```
Comment (asubramanya) [inline libgpopt/src/engine/CEngine.cpp]: s/Determing/Determine/

File: libgpopt/src/engine/CEngine.cpp [hunk 10]
```diff
@@ -2382,9 +2430,17 @@ CEngine::FCheckReqdPartPropagation
...
+//		This method is called after a group expression optimization job has
+//		started executing and can be used to cancel the job early.
+//
+//		This is usefule to prevent deadlocks when an enforcer optimizes same
+//		group with the same optimization context. Also, in case the subtree
+//		doesn't provided the required columns we can save optimization time by
```
Comment (asubramanya) [inline libgpopt/src/engine/CEngine.cpp]: s/provided/provide/

File: libgpopt/include/gpopt/base/CQueryContext.h (hunk 3 above)
Comment (hardikar) [reply libgpopt/include/gpopt/base/CQueryContext.h]: Added a note in `PqcGenerate()`. Thanks.

File: libgpopt/include/gpopt/search/CJobGroupExpressionOptimization.h (hunk 6 above)
Comment (hardikar) [reply libgpopt/include/gpopt/search/CJobGroupExpressionOptimization.h]: I think the statement is correct given the qualification "root by the current group expression". That is it finds the best child groups (which could include itself.

**Comment by hardikar** [comment]
Thank you very much all for reviewing!!

**Pull request merged**